src/relay/dnsrelay/mod.rs

```diff
@@ -0,0 +1,260 @@
+use std::{
+    future::Future,
+    net::{IpAddr, SocketAddr},
+    time::Duration,
+};
+
+use tokio::{
+    io::Result,
+    net::{TcpStream, UdpSocket},
+    prelude::*,
+    time,
+};
+
+use log::debug;
+
+use trust_dns_proto::{op::*, rr::*};
+
+use crate::{context::SharedContext, relay::socks5::*};
+
+pub async fn try_timeout<T, F>(fut: F, timeout: Option<Duration>) -> io::Result<T>
+where
+    F: Future<Output = Result<T>>,
+{
+    match timeout {
+        Some(t) => time::timeout(t, fut).await?,
+        None => fut.await,
+    }
+    .map_err(From::from)
+}
+
+async fn udp_lookup(qname: &Name, qtype: RecordType, server: SocketAddr) -> Result<Message> {
+    let mut socket = UdpSocket::bind(("0.0.0.0", 0)).await?;
+
+    let mut message = Message::new();
+    let mut query = Query::new();
+
+    query.set_query_type(qtype);
+    query.set_name(qname.clone());
+
+    message.set_id(6666);
```

Comment:
Message ID should be random, no? Same for below.

Reply:
Fixed.